core/os/flock/flock.go

```diff
@@ -15,229 +15,275 @@
 package flock
 
 import (
+	"bufio"
 	"errors"
+	"fmt"
+	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
 	"sync"
 	"time"
 )
 
-var errAlreadyLocked = errors.New("FLock already locked by others")
-var errCannotMakeDir = errors.New("Cannot make directory for FLock files")
-
 const (
-	fileSuffix = ".gapid-flock"
-	subDir     = "gapid-flocks"
+	fileSuffix   = ".gapid-flock"
+	realLockFile = "the-only-unique-truely-real-lock-file"
```

Comment:
The complexity level has risen significantly and I'd be really surprised if there wasn't a number of race conditions in this logic.

Let's go back to the original problems we're trying to solve:
* We want an inter-process locking mechanism so that a device is exclusively owned by a task for a period of time.
* We want the lock to be released if the process crashes. We should assume that the explicit unlock logic never gets an opportunity to be called, but the lock should automatically unlock in this case.

Doesn't the unix [`flock`](http://man7.org/linux/man-pages/man2/flock.2.html) and Windows [LockFile](https://msdn.microsoft.com/en-us/library/aa365202(v=vs.85).aspx) functions give you what you need here?

Reply:
But this is actually using the unix `flock` and windows `LockFile`. The timeout function is not supported by those sys calls.

Reply:
The complexity comes from the `timeout` mechanism. I need to make sure there is only a single process and a single thread when we modify the file (change the expiry time etc). So an extra flock is used to ensure that.

Let me see if I can simplify this a little bit more

Reply:
in 8d68cfe, the logic has been simplified.

Two lock files for each Mutex, one represents the device and has the expiry time written in it, another one is a RW lock, to guarantee only one thread/one process can remove the expired lock file which represents the device.